core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -261,14 +270,19 @@ private[spark] class ExecutorAllocationManager(
   private def updateAndSyncNumExecutorsTarget(now: Long): Int = synchronized {
     val maxNeeded = maxNumExecutorsNeeded
 
-    if (maxNeeded < numExecutorsTarget) {
+    if (initializing && (maxNeeded < numExecutorsTarget)) {
```

Comment:
the second part doesn't seem necessary. I don't think `maxNeeded` can be >= `numExecutorsTarget` if we are still initializing, since we only increase `maxNeeded` after submitting tasks.